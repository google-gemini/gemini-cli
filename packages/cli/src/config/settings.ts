/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import * as fs from 'fs';
import * as path from 'path';
import { homedir } from 'os';

export const SETTINGS_DIRECTORY_NAME = '.gemini';
export const USER_SETTINGS_DIR = path.join(homedir(), SETTINGS_DIRECTORY_NAME);
export const USER_SETTINGS_PATH = path.join(USER_SETTINGS_DIR, 'settings.json');

export enum SettingScope {
  User = 'User',
  Workspace = 'Workspace',
}

export interface Settings {
  theme?: string;
  sandbox?: boolean | string;
  toolDiscoveryCommand?: string;
  toolCallCommand?: string;
  mcpServerCommand?: string;
  // Add other settings here.
}

export interface SettingsFile {
  settings: Settings;
  path: string;
}
export class LoadedSettings {
  constructor(user: SettingsFile, workspace: SettingsFile) {
    this.user = user;
    this.workspace = workspace;
    this._merged = this.computeMergedSettings();
  }

  readonly user: SettingsFile;
  readonly workspace: SettingsFile;

  private _merged: Settings;

  get merged(): Settings {
    return this._merged;
  }

  private computeMergedSettings(): Settings {
    return {
      ...this.user.settings,
      ...this.workspace.settings,
    };
  }

  forScope(scope: SettingScope): SettingsFile {
    switch (scope) {
      case SettingScope.User:
        return this.user;
      case SettingScope.Workspace:
        return this.workspace;
      default:
        throw new Error(`Invalid scope: ${scope}`);
    }
  }

  setValue(
    scope: SettingScope,
    key: keyof Settings,
    value: string | undefined,
  ): void {
    const settingsFile = this.forScope(scope);
    settingsFile.settings[key] = value;
    this._merged = this.computeMergedSettings();
    saveSettings(settingsFile);
  }
}

/**
 * Loads settings from user and workspace directories.
 * Project settings override user settings.
 */
export function loadSettings(workspaceDir: string): LoadedSettings {
  let userSettings: Settings = {};
  let workspaceSettings = {};

  // Load user settings
  try {
    if (fs.existsSync(USER_SETTINGS_PATH)) {
      const userContent = fs.readFileSync(USER_SETTINGS_PATH, 'utf-8');
      userSettings = JSON.parse(userContent);
    }
  } catch (error) {
    console.error('Error reading user settings file:', error);
  }

  const workspaceSettingsPath = path.join(
    workspaceDir,
    SETTINGS_DIRECTORY_NAME,
    'settings.json',
  );

  // Load workspace settings
  try {
    if (fs.existsSync(workspaceSettingsPath)) {
      const projectContent = fs.readFileSync(workspaceSettingsPath, 'utf-8');
      workspaceSettings = JSON.parse(projectContent);
    }
  } catch (error) {
    console.error('Error reading workspace settings file:', error);
  }

  return new LoadedSettings(
    { path: USER_SETTINGS_PATH, settings: userSettings },
    { path: workspaceSettingsPath, settings: workspaceSettings },
  );
}

export function saveSettings(settingsFile: SettingsFile): void {
  try {
    // Ensure the directory exists
    const dirPath = path.dirname(settingsFile.path);
    if (!fs.existsSync(dirPath)) {
      fs.mkdirSync(dirPath, { recursive: true });
    }

    fs.writeFileSync(
      settingsFile.path,
      JSON.stringify(settingsFile.settings, null, 2),
      'utf-8',
    );
  } catch (error) {
    console.error('Error saving user settings file:', error);
  }
}
