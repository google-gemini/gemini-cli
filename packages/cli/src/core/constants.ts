export const MEMORY_FILE_NAME = 'GEMINI.md';