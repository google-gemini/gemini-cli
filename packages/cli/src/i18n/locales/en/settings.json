{
  "labels": {
    "theme": "Theme",
    "customThemes": "Custom Themes",
    "hideWindowTitle": "Hide Window Title",
    "hideTips": "Hide Tips",
    "hideBanner": "Hide Banner",
    "hideFooter": "Hide Footer",
    "showMemoryUsage": "Show Memory Usage",
    "language": "Language",
    "enableUsageStatistics": "Enable Usage Statistics",
    "autoConfigureMaxOldSpaceSize": "Auto Configure Max Old Space Size",
    "preferredEditor": "Preferred Editor",
    "maxSessionTurns": "Max Session Turns",
    "memoryImportFormat": "Memory Import Format",
    "memoryDiscoveryMaxDirs": "Memory Discovery Max Dirs",
    "contextFileName": "Context File Name",
    "vimMode": "Vim Mode",
    "ideMode": "IDE Mode",
    "accessibility": "Accessibility",
    "disableLoadingPhrases": "Disable Loading Phrases",
    "checkpointing": "Checkpointing",
    "enableCheckpointing": "Enable Checkpointing",
    "fileFiltering": "File Filtering",
    "enableRecursiveFileSearch": "Enable Recursive File Search",
    "disableAutoUpdate": "Disable Auto Update",
    "selectedAuthType": "Selected Auth Type",
    "useExternalAuth": "Use External Auth",
    "sandbox": "Sandbox",
    "coreTools": "Core Tools",
    "excludeTools": "Exclude Tools",
    "toolDiscoveryCommand": "Tool Discovery Command",
    "toolCallCommand": "Tool Call Command",
    "bugCommand": "Bug Command",
    "summarizeToolOutput": "Summarize Tool Output",
    "dnsResolutionOrder": "DNS Resolution Order",
    "excludedProjectEnvironmentVariables": "Excluded Project Environment Variables",
    "disableUpdateNag": "Disable Update Nag",
    "includeDirectories": "Include Directories",
    "loadMemoryFromIncludeDirectories": "Load Memory From Include Directories",
    "model": "Model",
    "hasSeenIdeIntegrationNudge": "Has Seen IDE Integration Nudge",
    "folderTrustFeature": "Folder Trust Feature",
    "folderTrust": "Folder Trust",
    "chatCompression": "Chat Compression",
    "showLineNumbers": "Show Line Numbers",
    "skipNextSpeakerCheck": "Disable Smart Conversation Flow",
    "respectGitIgnore": "Respect .gitignore",
    "respectGeminiIgnore": "Respect .geminiignore",
    "shouldUseNodePtyShell": "Use node-pty for Shell Execution"
  },
  "descriptions": {
    "theme": "The color theme for the UI.",
    "customThemes": "Custom theme definitions.",
    "hideWindowTitle": "Hide the window title bar",
    "hideTips": "Hide helpful tips in the UI",
    "hideBanner": "Hide the application banner",
    "hideFooter": "Hide the footer from the UI",
    "showMemoryUsage": "Display memory usage information in the UI",
    "language": "Interface language for the CLI",
    "enableUsageStatistics": "Enable collection of usage statistics",
    "autoConfigureMaxOldSpaceSize": "Automatically configure Node.js memory limits",
    "preferredEditor": "The preferred editor to open files in.",
    "maxSessionTurns": "Maximum number of user/model/tool turns to keep in a session. -1 means unlimited.",
    "memoryImportFormat": "The format to use when importing memory.",
    "memoryDiscoveryMaxDirs": "Maximum number of directories to search for memory.",
    "contextFileName": "The name of the context file.",
    "vimMode": "Enable Vim keybindings",
    "ideMode": "Enable IDE integration mode",
    "accessibility": "Accessibility settings.",
    "disableLoadingPhrases": "Disable loading phrases for accessibility",
    "checkpointing": "Session checkpointing settings.",
    "enableCheckpointing": "Enable session checkpointing for recovery",
    "fileFiltering": "Settings for git-aware file filtering.",
    "enableRecursiveFileSearch": "Enable recursive file search functionality",
    "disableAutoUpdate": "Disable automatic updates",
    "selectedAuthType": "The currently selected authentication type.",
    "useExternalAuth": "Whether to use an external authentication flow.",
    "sandbox": "Sandbox execution environment (can be a boolean or a path string).",
    "coreTools": "Paths to core tool definitions.",
    "excludeTools": "Tool names to exclude from discovery.",
    "toolDiscoveryCommand": "Command to run for tool discovery.",
    "toolCallCommand": "Command to run for tool calls.",
    "bugCommand": "Configuration for the bug report command.",
    "summarizeToolOutput": "Settings for summarizing tool output.",
    "dnsResolutionOrder": "The DNS resolution order.",
    "excludedProjectEnvironmentVariables": "Environment variables to exclude from project context.",
    "disableUpdateNag": "Disable update notification prompts.",
    "includeDirectories": "Additional directories to include in the workspace context.",
    "loadMemoryFromIncludeDirectories": "Whether to load memory files from include directories.",
    "model": "The Gemini model to use for conversations.",
    "hasSeenIdeIntegrationNudge": "Whether the user has seen the IDE integration nudge.",
    "folderTrustFeature": "Enable folder trust feature for enhanced security.",
    "folderTrust": "Setting to track whether Folder trust is enabled.",
    "chatCompression": "Chat compression settings.",
    "showLineNumbers": "Show line numbers in the chat.",
    "skipNextSpeakerCheck": "Disable AI-powered conversation flow analysis. When enabled, conversations will proceed using default behavior instead of intelligent analysis of who should speak next.",
    "respectGitIgnore": "Respect .gitignore files when searching.",
    "respectGeminiIgnore": "Respect .geminiignore files when searching.",
    "shouldUseNodePtyShell": "Use node-pty for shell command execution. Fallback to child_process still applies."
  },
  "languageOptions": {
    "english": "English",
    "spanish": "Español",
    "french": "Français"
  }
}
