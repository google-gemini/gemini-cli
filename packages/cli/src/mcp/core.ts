/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

// MCP Core stub - would be replaced with actual MCP SDK implementation
export interface McpServer {
  // Server interface methods would be defined here
}

export function createServer(options: any): McpServer {
  // Stub implementation
  return {} as McpServer;
}
