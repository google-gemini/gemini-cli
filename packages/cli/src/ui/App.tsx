/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { useCallback, useEffect, useMemo, useState } from 'react';
import { Box, Static, Text, useStdout } from 'ink';
import { StreamingState, type HistoryItem } from './types.js';
import { useGeminiStream } from './hooks/useGeminiStream.js';
import { useLoadingIndicator } from './hooks/useLoadingIndicator.js';
import { useThemeCommand } from './hooks/useThemeCommand.js';
import { useSlashCommandProcessor } from './hooks/slashCommandProcessor.js';
import { Header } from './components/Header.js';
import { LoadingIndicator } from './components/LoadingIndicator.js';
import { EditorState, InputPrompt } from './components/InputPrompt.js';
import { Footer } from './components/Footer.js';
import { ThemeDialog } from './components/ThemeDialog.js';
import { shortenPath, type Config } from '@gemini-code/server';
import { Colors } from './colors.js';
import { Help } from './components/Help.js';
import { loadHierarchicalGeminiMemory } from '../config/config.js';
import { LoadedSettings } from '../config/settings.js';
import { Tips } from './components/Tips.js';
import { ConsoleOutput } from './components/ConsolePatcher.js';
import { HistoryItemDisplay } from './components/HistoryItemDisplay.js';
import { useCompletion } from './hooks/useCompletion.js';
import { SuggestionsDisplay } from './components/SuggestionsDisplay.js';
import { isAtCommand, isSlashCommand } from './utils/commandUtils.js';
import { useHistory } from './hooks/useHistoryManager.js';
import process from 'node:process'; // For performMemoryRefresh
import { MessageType } from './types.js'; // For performMemoryRefresh
import { getErrorMessage } from '@gemini-code/server'; // For performMemoryRefresh

interface AppProps {
  config: Config;
  settings: LoadedSettings;
  cliVersion: string;
  startupWarnings?: string[];
}

export const App = ({
  config,
  settings,
  cliVersion,
  startupWarnings = [],
}: AppProps) => {
  const { history, addItem, clearItems } = useHistory();
  const [staticKey, setStaticKey] = useState(0);
  const refreshStatic = useCallback(() => {
    setStaticKey((prev) => prev + 1);
  }, [setStaticKey]);

  const [geminiMdFileCount, setGeminiMdFileCount] = useState<number>(0); // Added for memory file count
  const [debugMessage, setDebugMessage] = useState<string>('');
  const [showHelp, setShowHelp] = useState<boolean>(false);
  const [themeError, setThemeError] = useState<string | null>(null);

  const {
    isThemeDialogOpen,
    openThemeDialog,
    handleThemeSelect,
    handleThemeHighlight,
  } = useThemeCommand(settings, setThemeError);

  // useEffect to initialize geminiMdFileCount from config when config is ready
  useEffect(() => {
    if (config) {
      setGeminiMdFileCount(config.getGeminiMdFileCount());
    }
  }, [config]);

  const performMemoryRefresh = useCallback(async () => {
    addItem(
      {
        type: MessageType.INFO,
        text: 'Refreshing hierarchical memory (GEMINI.md files)...',
      },
      Date.now(),
    );
    try {
      const { memoryContent, fileCount } = await loadHierarchicalGeminiMemory(
        process.cwd(),
        config.getDebugMode(),
      );
      config.setUserMemory(memoryContent);
      config.setGeminiMdFileCount(fileCount);
      setGeminiMdFileCount(fileCount);

      // chatSessionRef.current = null; // This was in useGeminiStream, might need similar logic or pass chat ref
      addItem(
        {
          type: MessageType.INFO,
          text: `Memory refreshed successfully. ${memoryContent.length > 0 ? `Loaded ${memoryContent.length} characters from ${fileCount} file(s).` : 'No memory content found.'}`,
        },
        Date.now(),
      );
      if (config.getDebugMode()) {
        console.log(
          `[DEBUG] Refreshed memory content in config: ${memoryContent.substring(0, 200)}...`,
        );
      }
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      addItem(
        {
          type: MessageType.ERROR,
          text: `Error refreshing memory: ${errorMessage}`,
        },
        Date.now(),
      );
      console.error('Error refreshing memory:', error);
    }
  }, [config, addItem]);

  const { handleSlashCommand, slashCommands } = useSlashCommandProcessor(
    config, // Pass config
    addItem,
    clearItems,
    refreshStatic,
    setShowHelp,
    setDebugMessage,
    openThemeDialog,
    performMemoryRefresh,
  );

  const { streamingState, submitQuery, initError, pendingHistoryItem } =
    useGeminiStream(
      addItem,
      clearItems,
      refreshStatic,
      setShowHelp,
      config,
      setDebugMessage,
      openThemeDialog,
      handleSlashCommand,
    );
  const { elapsedTime, currentLoadingPhrase } =
    useLoadingIndicator(streamingState);

  const handleFinalSubmit = useCallback(
    (submittedValue: string) => {
      const trimmedValue = submittedValue.trim();
      if (trimmedValue.length > 0) {
        submitQuery(submittedValue);
      }
    },
    [submitQuery],
  );

  const userMessages = useMemo(
    () =>
      history
        .filter(
          (item): item is HistoryItem & { type: 'user'; text: string } =>
            item.type === 'user' &&
            typeof item.text === 'string' &&
            item.text.trim() !== '',
        )
        .map((item) => item.text),
    [history],
  );

  const isInputActive = streamingState === StreamingState.Idle && !initError;

  const [query, setQuery] = useState('');
  const [editorState, setEditorState] = useState<EditorState>({
    key: 0,
    initialCursorOffset: undefined,
  });

  const onChangeAndMoveCursor = useCallback(
    (value: string) => {
      setQuery(value);
      setEditorState((s) => ({
        key: s.key + 1,
        initialCursorOffset: value.length,
      }));
    },
    [setQuery, setEditorState],
  );

  const handleClearScreen = useCallback(() => {
    clearItems();
    console.clear();
    refreshStatic();
  }, [clearItems, refreshStatic]);

  const completion = useCompletion(
    query,
    config.getTargetDir(),
    isInputActive && (isAtCommand(query) || isSlashCommand(query)),
    slashCommands,
  );

  // --- Render Logic ---

  // Get terminal width
  const { stdout } = useStdout();
  const terminalWidth = stdout?.columns ?? 80;
  // Calculate width for suggestions, leave some padding
  const suggestionsWidth = Math.max(60, Math.floor(terminalWidth * 0.8));

  return (
    <Box flexDirection="column" marginBottom={1} width="90%">
      {/*
       * The Static component is an Ink intrinsic in which there can only be 1 per application.
       * Because of this restriction we're hacking it slightly by having a 'header' item here to
       * ensure that it's statically rendered.
       *
       * Background on the Static Item: Anything in the Static component is written a single time
       * to the console. Think of it like doing a console.log and then never using ANSI codes to
       * clear that content ever again. Effectively it has a moving frame that every time new static
       * content is set it'll flush content to the terminal and move the area which it's "clearing"
       * down a notch. Without Static the area which gets erased and redrawn continuously grows.
       */}
      <Static key={'static-key-' + staticKey} items={['header', ...history]}>
        {(item, index) => {
          if (item === 'header') {
            return (
              <Box flexDirection="column" key={'header-' + index}>
                <Header />
                <Tips />
              </Box>
            );
          }

          const historyItem = item as HistoryItem;
          return (
            <HistoryItemDisplay
              key={'history-' + historyItem.id}
              item={historyItem}
              onSubmit={submitQuery}
            />
          );
        }}
      </Static>
      {pendingHistoryItem && (
        <HistoryItemDisplay
          // TODO(taehykim): It seems like references to ids aren't necessary in
          // HistoryItemDisplay. Refactor later. Use a fake id for now.
          item={{ ...pendingHistoryItem, id: 0 }}
          onSubmit={submitQuery}
        />
      )}
      {showHelp && <Help commands={slashCommands} />}

      {startupWarnings.length > 0 && (
        <Box
          borderStyle="round"
          borderColor={Colors.AccentYellow}
          paddingX={1}
          marginY={1}
          flexDirection="column"
        >
          {startupWarnings.map((warning, index) => (
            <Text key={index} color={Colors.AccentYellow}>
              {warning}
            </Text>
          ))}
        </Box>
      )}

      {isThemeDialogOpen ? (
        <Box flexDirection="column">
          {themeError && (
            <Box marginBottom={1}>
              <Text color={Colors.AccentRed}>{themeError}</Text>
            </Box>
          )}
          <ThemeDialog
            onSelect={handleThemeSelect}
            onHighlight={handleThemeHighlight}
            settings={settings}
            setQuery={setQuery}
          />
        </Box>
      ) : (
        <>
          <LoadingIndicator
            isLoading={streamingState === StreamingState.Responding}
            currentLoadingPhrase={currentLoadingPhrase}
            elapsedTime={elapsedTime}
          />
          {isInputActive && (
            <>
              <Box
                marginTop={1}
                display="flex"
                justifyContent="space-between"
                width="100%"
              >
                <Box>
                  <Text color={Colors.SubtleComment}>cwd: </Text>
                  <Text color={Colors.LightBlue}>
                    {shortenPath(config.getTargetDir(), 70)}
                  </Text>
                </Box>
              </Box>

              <InputPrompt
                query={query}
                onChange={setQuery}
                onChangeAndMoveCursor={onChangeAndMoveCursor}
                editorState={editorState}
                onSubmit={handleFinalSubmit} // Pass handleFinalSubmit directly
                showSuggestions={completion.showSuggestions}
                suggestions={completion.suggestions}
                activeSuggestionIndex={completion.activeSuggestionIndex}
                userMessages={userMessages} // Pass userMessages
                navigateSuggestionUp={completion.navigateUp}
                navigateSuggestionDown={completion.navigateDown}
                resetCompletion={completion.resetCompletionState}
                setEditorState={setEditorState}
                onClearScreen={handleClearScreen} // Added onClearScreen prop
              />
              {completion.showSuggestions && (
                <Box>
                  <SuggestionsDisplay
                    suggestions={completion.suggestions}
                    activeIndex={completion.activeSuggestionIndex}
                    isLoading={completion.isLoadingSuggestions}
                    width={suggestionsWidth}
                    scrollOffset={completion.visibleStartIndex}
                  />
                </Box>
              )}
            </>
          )}
        </>
      )}

      {initError && streamingState !== StreamingState.Responding && (
        <Box
          borderStyle="round"
          borderColor={Colors.AccentRed}
          paddingX={1}
          marginBottom={1}
        >
          {history.find(
            (item) => item.type === 'error' && item.text?.includes(initError),
          )?.text ? (
            <Text color={Colors.AccentRed}>
              {
                history.find(
                  (item) =>
                    item.type === 'error' && item.text?.includes(initError),
                )?.text
              }
            </Text>
          ) : (
            <>
              <Text color={Colors.AccentRed}>
                Initialization Error: {initError}
              </Text>
              <Text color={Colors.AccentRed}>
                {' '}
                Please check API key and configuration.
              </Text>
            </>
          )}
        </Box>
      )}

      <Footer
        config={config}
        debugMode={config.getDebugMode()}
        debugMessage={debugMessage}
        cliVersion={cliVersion}
        geminiMdFileCount={geminiMdFileCount}
      />
      <ConsoleOutput />
    </Box>
  );
};
