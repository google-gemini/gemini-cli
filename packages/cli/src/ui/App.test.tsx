/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { describe, it, expect, vi, beforeEach, afterEach, Mock } from 'vitest';
import { render } from 'ink-testing-library';
import { App } from './App.js';
import { Config as ServerConfig, MCPServerConfig } from '@gemini-code/core';
import type { ToolRegistry } from '@gemini-code/core';
import { LoadedSettings, SettingsFile, Settings } from '../config/settings.js';

// Define a more complete mock server config based on actual Config
interface MockServerConfig {
  apiKey: string;
  model: string;
  sandbox: boolean | string;
  targetDir: string;
  debugMode: boolean;
  question?: string;
  fullContext: boolean;
  coreTools?: string[];
  toolDiscoveryCommand?: string;
  toolCallCommand?: string;
  mcpServerCommand?: string;
  mcpServers?: Record<string, MCPServerConfig>; // Use imported MCPServerConfig
  userAgent: string;
  userMemory: string;
  geminiMdFileCount: number;
  alwaysSkipModificationConfirmation: boolean;
  vertexai?: boolean;
  showMemoryUsage?: boolean;

  getApiKey: Mock<() => string>;
  getModel: Mock<() => string>;
  getSandbox: Mock<() => boolean | string>;
  getTargetDir: Mock<() => string>;
  getToolRegistry: Mock<() => ToolRegistry>; // Use imported ToolRegistry type
  getDebugMode: Mock<() => boolean>;
  getQuestion: Mock<() => string | undefined>;
  getFullContext: Mock<() => boolean>;
  getCoreTools: Mock<() => string[] | undefined>;
  getToolDiscoveryCommand: Mock<() => string | undefined>;
  getToolCallCommand: Mock<() => string | undefined>;
  getMcpServerCommand: Mock<() => string | undefined>;
  getMcpServers: Mock<() => Record<string, MCPServerConfig> | undefined>;
  getUserAgent: Mock<() => string>;
  getUserMemory: Mock<() => string>;
  setUserMemory: Mock<(newUserMemory: string) => void>;
  getGeminiMdFileCount: Mock<() => number>;
  setGeminiMdFileCount: Mock<(count: number) => void>;
  getAlwaysSkipModificationConfirmation: Mock<() => boolean>;
  setAlwaysSkipModificationConfirmation: Mock<(skip: boolean) => void>;
  getVertexAI: Mock<() => boolean | undefined>;
  getShowMemoryUsage: Mock<() => boolean>;
}

// Mock @gemini-code/core and its Config class
vi.mock('@gemini-code/core', async (importOriginal) => {
  const actualCore = await importOriginal<typeof import('@gemini-code/core')>();
  const ConfigClassMock = vi
    .fn()
    .mockImplementation((optionsPassedToConstructor) => {
      const opts = { ...optionsPassedToConstructor }; // Clone
      // Basic mock structure, will be extended by the instance in tests
      return {
        apiKey: opts.apiKey || 'test-key',
        model: opts.model || 'test-model-in-mock-factory',
        sandbox: typeof opts.sandbox === 'boolean' ? opts.sandbox : false,
        targetDir: opts.targetDir || '/test/dir',
        debugMode: opts.debugMode || false,
        question: opts.question,
        fullContext: opts.fullContext ?? false,
        coreTools: opts.coreTools,
        toolDiscoveryCommand: opts.toolDiscoveryCommand,
        toolCallCommand: opts.toolCallCommand,
        mcpServerCommand: opts.mcpServerCommand,
        mcpServers: opts.mcpServers,
        userAgent: opts.userAgent || 'test-agent',
        userMemory: opts.userMemory || '',
        geminiMdFileCount: opts.geminiMdFileCount || 0,
        alwaysSkipModificationConfirmation:
          opts.alwaysSkipModificationConfirmation ?? false,
        vertexai: opts.vertexai,
        showMemoryUsage: opts.showMemoryUsage ?? false,

        getApiKey: vi.fn(() => opts.apiKey || 'test-key'),
        getModel: vi.fn(() => opts.model || 'test-model-in-mock-factory'),
        getSandbox: vi.fn(() =>
          typeof opts.sandbox === 'boolean' ? opts.sandbox : false,
        ),
        getTargetDir: vi.fn(() => opts.targetDir || '/test/dir'),
        getToolRegistry: vi.fn(() => ({}) as ToolRegistry), // Simple mock
        getDebugMode: vi.fn(() => opts.debugMode || false),
        getQuestion: vi.fn(() => opts.question),
        getFullContext: vi.fn(() => opts.fullContext ?? false),
        getCoreTools: vi.fn(() => opts.coreTools),
        getToolDiscoveryCommand: vi.fn(() => opts.toolDiscoveryCommand),
        getToolCallCommand: vi.fn(() => opts.toolCallCommand),
        getMcpServerCommand: vi.fn(() => opts.mcpServerCommand),
        getMcpServers: vi.fn(() => opts.mcpServers),
        getUserAgent: vi.fn(() => opts.userAgent || 'test-agent'),
        getUserMemory: vi.fn(() => opts.userMemory || ''),
        setUserMemory: vi.fn(),
        getGeminiMdFileCount: vi.fn(() => opts.geminiMdFileCount || 0),
        setGeminiMdFileCount: vi.fn(),
        getAlwaysSkipModificationConfirmation: vi.fn(
          () => opts.alwaysSkipModificationConfirmation ?? false,
        ),
        setAlwaysSkipModificationConfirmation: vi.fn(),
        getVertexAI: vi.fn(() => opts.vertexai),
        getShowMemoryUsage: vi.fn(() => opts.showMemoryUsage ?? false),
      };
    });
  return {
    ...actualCore,
    Config: ConfigClassMock,
    MCPServerConfig: actualCore.MCPServerConfig,
  };
});

// Mock heavy dependencies or those with side effects
vi.mock('./hooks/useGeminiStream', () => ({
  useGeminiStream: vi.fn(() => ({
    streamingState: 'Idle',
    submitQuery: vi.fn(),
    initError: null,
    pendingHistoryItems: [],
  })),
}));

vi.mock('./hooks/useLogger', () => ({
  useLogger: vi.fn(() => ({
    getPreviousUserMessages: vi.fn().mockResolvedValue([]),
  })),
}));

vi.mock('../config/config.js', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    // @ts-expect-error - this is fine
    ...actual,
    loadHierarchicalGeminiMemory: vi
      .fn()
      .mockResolvedValue({ memoryContent: '', fileCount: 0 }),
  };
});

describe('App UI', () => {
  let mockConfig: MockServerConfig;
  let mockSettings: LoadedSettings;
  let currentUnmount: (() => void) | undefined;

  const createMockSettings = (
    settings: Partial<Settings> = {},
  ): LoadedSettings => {
    const userSettingsFile: SettingsFile = {
      path: '/user/settings.json',
      settings: {},
    };
    const workspaceSettingsFile: SettingsFile = {
      path: '/workspace/.gemini/settings.json',
      settings,
    };
    return new LoadedSettings(userSettingsFile, workspaceSettingsFile);
  };

  beforeEach(() => {
    const ServerConfigMocked = vi.mocked(ServerConfig, true);
    mockConfig = new ServerConfigMocked({
      apiKey: 'test-key',
      model: 'test-model-in-options',
      sandbox: false,
      targetDir: '/test/dir',
      debugMode: false,
      userAgent: 'test-agent',
      userMemory: '',
      geminiMdFileCount: 0,
      showMemoryUsage: false,
      // Provide other required fields for ConfigParameters if necessary
    }) as unknown as MockServerConfig;

    // Ensure the getShowMemoryUsage mock function is specifically set up if not covered by constructor mock
    if (!mockConfig.getShowMemoryUsage) {
      mockConfig.getShowMemoryUsage = vi.fn(() => false);
    }
    mockConfig.getShowMemoryUsage.mockReturnValue(false); // Default for most tests

    mockSettings = createMockSettings();
  });

  afterEach(() => {
    if (currentUnmount) {
      currentUnmount();
      currentUnmount = undefined;
    }
    vi.clearAllMocks(); // Clear mocks after each test
  });

  it('should display default "GEMINI.md" in footer when contextFileName is not set and count is 1', async () => {
    mockConfig.getGeminiMdFileCount.mockReturnValue(1);
    // For this test, ensure showMemoryUsage is false or debugMode is false if it relies on that
    mockConfig.getDebugMode.mockReturnValue(false);
    mockConfig.getShowMemoryUsage.mockReturnValue(false);

    const { lastFrame, unmount } = render(
      <App
        config={mockConfig as unknown as ServerConfig}
        settings={mockSettings}
        cliVersion="1.0.0"
      />,
    );
    currentUnmount = unmount;
    await Promise.resolve(); // Wait for any async updates
    expect(lastFrame()).toContain('Using 1 GEMINI.md file');
  });

  it('should display default "GEMINI.md" with plural when contextFileName is not set and count is > 1', async () => {
    mockConfig.getGeminiMdFileCount.mockReturnValue(2);
    mockConfig.getDebugMode.mockReturnValue(false);
    mockConfig.getShowMemoryUsage.mockReturnValue(false);

    const { lastFrame, unmount } = render(
      <App
        config={mockConfig as unknown as ServerConfig}
        settings={mockSettings}
        cliVersion="1.0.0"
      />,
    );
    currentUnmount = unmount;
    await Promise.resolve();
    expect(lastFrame()).toContain('Using 2 GEMINI.md files');
  });

  it('should display custom contextFileName in footer when set and count is 1', async () => {
    mockSettings = createMockSettings({ contextFileName: 'AGENTS.MD' });
    mockConfig.getGeminiMdFileCount.mockReturnValue(1);
    mockConfig.getDebugMode.mockReturnValue(false);
    mockConfig.getShowMemoryUsage.mockReturnValue(false);

    const { lastFrame, unmount } = render(
      <App
        config={mockConfig as unknown as ServerConfig}
        settings={mockSettings}
        cliVersion="1.0.0"
      />,
    );
    currentUnmount = unmount;
    await Promise.resolve();
    expect(lastFrame()).toContain('Using 1 AGENTS.MD file');
  });

  it('should display custom contextFileName with plural when set and count is > 1', async () => {
    mockSettings = createMockSettings({ contextFileName: 'MY_NOTES.TXT' });
    mockConfig.getGeminiMdFileCount.mockReturnValue(3);
    mockConfig.getDebugMode.mockReturnValue(false);
    mockConfig.getShowMemoryUsage.mockReturnValue(false);

    const { lastFrame, unmount } = render(
      <App
        config={mockConfig as unknown as ServerConfig}
        settings={mockSettings}
        cliVersion="1.0.0"
      />,
    );
    currentUnmount = unmount;
    await Promise.resolve();
    expect(lastFrame()).toContain('Using 3 MY_NOTES.TXT files');
  });

  it('should not display context file message if count is 0, even if contextFileName is set', async () => {
    mockSettings = createMockSettings({ contextFileName: 'ANY_FILE.MD' });
    mockConfig.getGeminiMdFileCount.mockReturnValue(0);
    mockConfig.getDebugMode.mockReturnValue(false);
    mockConfig.getShowMemoryUsage.mockReturnValue(false);

    const { lastFrame, unmount } = render(
      <App
        config={mockConfig as unknown as ServerConfig}
        settings={mockSettings}
        cliVersion="1.0.0"
      />,
    );
    currentUnmount = unmount;
    await Promise.resolve();
    expect(lastFrame()).not.toContain('ANY_FILE.MD');
  });
});
