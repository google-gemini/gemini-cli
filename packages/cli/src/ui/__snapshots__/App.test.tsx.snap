// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App UI > should render correctly with the prompt input box 1`] = `
"
                                                   queue: wait_for_idle (Ctrl+M to toggle)
╭────────────────────────────────────────────────────────────────────────────────────────╮
│ >   Type your message or @path/to/file                                                 │
╰────────────────────────────────────────────────────────────────────────────────────────╯
/test/dir               no sandbox (see /docs)                 model (100% context left)"
`;

exports[`App UI > should render the initial UI correctly 1`] = `
" I'm Feeling Lucky (esc to cancel, 0s)

                                                   queue: wait_for_idle (Ctrl+M to toggle)
/test/dir               no sandbox (see /docs)                 model (100% context left)"
`;

exports[`App UI > when in a narrow terminal > should render with a column layout 1`] = `
"


  queue: wait_for_idle (Ctrl+M to toggle)
╭────────────────────────────────────────────────────────────────────────────────────────╮
│ >   Type your message or @path/to/file                                                 │
╰────────────────────────────────────────────────────────────────────────────────────────╯
dir

no sandbox (see /docs)

model (100% context left)  | ✖ 5 errors (ctrl+o for details)"
`;
